applications/RomApplication/python_scripts/petrov_galerkin_training_utility.py

```diff
@@ -0,0 +1,158 @@
+        # Note that this ensures that the residuals utility is created in the first residuals append call
+        # If not, it might happen that the solver scheme is created by the ROM residuals call rather than by the solver one
+        if not hasattr(self, '__rom_residuals_utility'):
+            self.__rom_residuals_utility = KratosROM.RomResidualsUtility(
+                computing_model_part,
+                self.rom_settings,
+                self.solver._GetScheme())
+
+            if self.echo_level > 0 : KratosMultiphysics.Logger.PrintInfo("PetrovGalerkinTrainingUtility","RomResidualsUtility created.")
+
+        # Generate the matrix of residuals or projected Jacobians.
+        if self.basis_strategy=="Jacobian":
+            snapshots_matrix = self.__rom_residuals_utility.GetProjectedGlobalLHS()
+            if self.echo_level > 0 : KratosMultiphysics.Logger.PrintInfo("PetrovGalerkinTrainingUtility","Generated matrix of projected Jacobian.")
+        elif self.basis_strategy=="Residuals":
+            snapshots_matrix = []
+            files_to_read_and_delete = glob('*.res.mm')#TODO: Stop writing to disk.
+            for to_erase_file in files_to_read_and_delete:
+                non_converged_iteration_snapshot = KratosMultiphysics.Vector()
+                KratosMultiphysics.ReadMatrixMarketVector(to_erase_file, non_converged_iteration_snapshot)
+                snapshots_matrix.append(non_converged_iteration_snapshot)
+            snapshots_matrix = np.array(snapshots_matrix).T
+            if self.echo_level > 0 : KratosMultiphysics.Logger.PrintInfo("PetrovGalerkinTrainingUtility","Generating matrix of residuals.")
+        else:
+            err_msg = "\'self.basis_strategy\' is not available. Select either Jacobian or Residuals."
+            raise Exception(err_msg)
+
+        np_snapshots_matrix = np.array(snapshots_matrix, copy=False)
+        self.time_step_snapshots_matrix_container.append(np_snapshots_matrix)
+
+    def CalculateAndSaveBasis(self):
+        # Calculate the new basis and save
+        snapshots_basis = self.__CalculateResidualBasis()
+        self.__AppendNewBasisToRomParameters(snapshots_basis)
+
+
+    @classmethod
+    def __GetPetrovGalerkinTrainingDefaultSettings(cls):
+        default_settings = KratosMultiphysics.Parameters("""{
+                "train": false,
+                "basis_strategy": "Residuals",
+                "include_phi": false,
+                "svd_truncation_tolerance": 1.0e-6,
+                "echo_level": 0
+        }""")
+        return default_settings
+
+    def __CalculateResidualBasis(self):
+        # Set up the snapshots matrix for new basis
+        n_steps = len(self.time_step_snapshots_matrix_container)
+        snapshots_matrix = self.time_step_snapshots_matrix_container[0]
+        for i in range(1,n_steps):
+            del self.time_step_snapshots_matrix_container[0] # Avoid having two matrices, numpy does not concatenate references.
```

Comment:
We can try and use something like `np.block(my_list)`. This transforms a python list contaning numpy arrays into a single array. 
It might be memory demanding though

Reply:
This part is to avoid having one list meanwhile building the numpy array, even with np.block() you will duplicate the data when creating the new array. I will write a note for future check.